src/main.rs

```diff
@@ -1,3 +1,53 @@
-fn main() {
-    println!("Hello, world!");
+use std::sync::Arc;
+
+use clap::Parser;
+use ethers::providers::{Http, Middleware, ProviderError, StreamExt};
+use ethers::signers::LocalWallet;
+use ethers::types::{Block, Transaction};
+
+use crate::cli::{ClientParseError, Opts};
+use crate::client::Client;
+
+pub mod cli;
+pub mod client;
+
+/// Represents global errors emitted from the bot process itself
+#[derive(Debug)]
+pub enum Error {
+    /// Error parsing client from provided arguments
+    ClientOptsError(ClientParseError),
+    /// Error in blockchain interaction
+    ProviderError(ProviderError),
+}
+
+impl From<ClientParseError> for Error {
+    fn from(value: ClientParseError) -> Self {
+        Self::ClientOptsError(value)
+    }
+}
+
+impl From<ProviderError> for Error {
+    fn from(value: ProviderError) -> Self {
+        Self::ProviderError(value)
+    }
+}
+
+#[tokio::main]
+async fn main() -> Result<(), Error> {
+    let opts: Opts = Opts::parse();
+    let client: Arc<Client<LocalWallet, Http>> = Arc::new(opts.try_into()?);
+
+    let mut stream = client.provider().watch_blocks().await?;
+
+    /* listen for new blocks */
+    while let Some(block_hash) = stream.next().await {
+        let block: Block<Transaction> = client
+            .provider()
+            .get_block_with_txs(block_hash)
+            .await?
+            .unwrap();
+        println!("{}", serde_json::to_string(&block).expect("Invalid data"));
```

Comment:
is this the same as writing to standard output? It stood out to me and I ended up finding this

https://github.com/BurntSushi/advent-of-code/issues/17

which makes a case for using `writeln!(io::stdout, "...")` instead of `println!()`

Reply:
In the same way a distinction exists in C between `fprintf(stdout, "Hello, world!")` and `printf("Hello, world!")` I suppose?

In reading the linked issue, I have to admit, I didn't know `println!` would panic on error! For what it's worth, [Lighthouse](https://github.com/sigp/lighthouse/search?q=println) doesn't seem to mind? Still good to know that `println!` panics.

Reply:
@ScaredIbis Actually I agree with you here.

Reply:
yep fair enough, my assumption after looking at this code is that this will be a building block within a larger system where it would be desirable for errors to be handled in whatever is consuming this stream of blocks

Reply:
Consult ff1dc3e.